packages/core/src/transactions/utils/transactions.ts

```diff
@@ -1,212 +1,311 @@
+
+const erc20ABI = [
+  'function balanceOf(address _who) public view returns (uint256)',
+  'function allowance(address _owner, address _spender) public view returns (uint256)',
+  'function approve(address _spender, uint256 _value) public returns (bool)',
+]
+
+const forwarderAbi = [
+  'function forward(bytes evmCallScript) public',
+  'function canForward(address sender, bytes evmCallScript) public view returns (bool)',
+  'function isForwarder() external pure returns (bool)',
+]
+
+const forwarderFeeAbi = [
+  'function forwardFee() external view returns (address, uint256)',
+]
+
+// Is the given method a full signature, e.g. 'foo(arg1,arg2,...)'
+export const isFullMethodSignature = (methodSignature: string): boolean => {
+  return (
+    Boolean(methodSignature) &&
+    methodSignature.includes('(') &&
+    methodSignature.includes(')')
+  )
+}
+
+export interface directTransaction extends TransactionRequest {
+  token?: {
+    address: string
+    value: string
+    spender: string
+  }
+  pretransaction?: {
+    from: string
+    to: string
+    data: string
+  }
+}
+
+export async function applyPretransaction(
+  directTransaction: directTransaction
+) {
+  // Token allowance pretransaction
+  if (directTransaction.token) {
+    const {
+      from,
+      to,
+      token: { address: tokenAddress, value: tokenValue, spender },
+    } = directTransaction
+
+    // Approve the transaction destination unless an spender is passed to approve a different contract
+    const approveSpender = spender || to
+
+    // TODO: handle provider network: using chainId from Transaction Request?
+    const provider = ethers.getDefaultProvider()
+
+    const tokenContract = new ethers.Contract(tokenAddress, erc20ABI, provider)
+    const balance = await tokenContract.balanceOf(from)
+
+    const tokenValueBN = ethers.utils.bigNumberify(tokenValue)
+
+    if (ethers.utils.bigNumberify(balance).lt(tokenValueBN)) {
+      throw new Error(
+        `Balance too low. ${from} balance of ${tokenAddress} token is ${balance} (attempting to send ${tokenValue})`
+      )
+    }
+
+    const allowance = await tokenContract.allowance(from, approveSpender)
+    const allowanceBN = ethers.utils.bigNumberify(allowance)
+
+    // If allowance is already greater than or equal to amount, there is no need to do an approve transaction
+    if (allowanceBN.lt(tokenValueBN)) {
+      if (allowanceBN.gt(ethers.constants.Zero)) {
+        // TODO: Actually handle existing approvals (some tokens fail when the current allowance is not 0)
+        console.warn(
+          `${from} already approved ${approveSpender}. In some tokens, approval will fail unless the allowance is reset to 0 before re-approving again.`
+        )
+      }
+
+      const tokenApproveTransaction = {
+        // TODO: should we include transaction options?
+        from,
+        to: tokenAddress,
+        // TODO: verify how to encode the function call
+        data: ethers.utils.RLP.encode(
```

Comment:
Still can't figure out how this encoding is done on ethers. We were doing: `tokenContract.methods.approve(approveSpender, tokenValue).encodeABI()` on web3.